sympy/solvers/ode/tests/test_ode.py

```diff
@@ -565,22 +567,14 @@ def test_dsolve_linsystem_symbol_piecewise():
     u = Symbol('u')  # XXX it's more complicated with real u
     eq = (Eq(diff(f(x), x), 2*f(x) + g(x)),
            Eq(diff(g(x), x), u*f(x)))
-    s1 = [Eq(f(x), Piecewise((C1*exp(x*(sqrt(4*u + 4)/2 + 1)) +
-        C2*exp(x*(-sqrt(4*u + 4)/2 + 1)), Ne(4*u + 4, 0)), ((C1 + C2*(x +
-        Piecewise((0, Eq(sqrt(4*u + 4)/2 + 1, 2)), (1/(-sqrt(4*u + 4)/2 + 1),
-        True))))*exp(x*(sqrt(4*u + 4)/2 + 1)), True))), Eq(g(x),
-        Piecewise((C1*(sqrt(4*u + 4)/2 - 1)*exp(x*(sqrt(4*u + 4)/2 + 1)) +
-        C2*(-sqrt(4*u + 4)/2 - 1)*exp(x*(-sqrt(4*u + 4)/2 + 1)), Ne(4*u + 4,
-        0)), ((C1*(sqrt(4*u + 4)/2 - 1) + C2*(x*(sqrt(4*u + 4)/2 - 1) +
-        Piecewise((1, Eq(sqrt(4*u + 4)/2 + 1, 2)), (0,
-        True))))*exp(x*(sqrt(4*u + 4)/2 + 1)), True)))]
+    s1 = [
+        Eq(f(x), -C1*exp(x*(1 - sqrt(u + 1)))/(sqrt(u + 1) + 1) + C2*exp(x*(sqrt(u + 1) + 1))/(sqrt(u + 1) - 1)),
+        Eq(g(x), C1*exp(x*(1 - sqrt(u + 1))) + C2*exp(x*(sqrt(u + 1) + 1)))
+    ]
+
     assert dsolve(eq) == s1
-    # FIXME: assert checksysodesol(eq, s) == (True, [0, 0])
-    # Remove lines below when checksysodesol works
-    s = [(l.lhs, l.rhs) for l in s1]
-    for v in [0, 7, -42, 5*I, 3 + 4*I]:
-        assert eq[0].subs(s).subs(u, v).doit().simplify()
-        assert eq[1].subs(s).subs(u, v).doit().simplify()
+    # assert checksysodesol(eq, s) == (True, [0, 0]) is fixed
+    assert checksysodesol(eq, s1) == (True, [0, 0])
```

Comment:
The solution here has changed. Was the original solution also correct? If checksysodesol *should* have worked then we should open an issue or an XFAIL test for that.

Reply:
Well this solution is correct as well and this solution passed the checksysodesol test. But the main reason why so many solutions like this have to be changed is because of the minor things like constant number and that solution will differ by factor, for example one method may give a solution:
```
C1 * t**2 + C2 * t / 2
```
And one method may give this solution:
```
2 * C1 * t**2 + C2 * t
```
Both the solution are effectively similar with just a difference of arbitrary constants. So these kind of conflicts were quite common. Hence I verified if the new solution passed checksysodesol and according to my knowledge, all of new ones and everything did.
```

Reply:
That's all fine. The question is that if the old solution is correct but fails checksysodesol then that's a bug in checksysodesol. It is fine to change this test but there should be a new XFAIL tests or an issue opened for the example that fails checksysodesol.

Reply:
@oscarbenjamin When I tried to solve the equation manually, it became clear that the solution provided by the original author of the test case is correct. There are namely two cases where the solution provided by our code will fail and we will have to look into it. Along with that I have found some limitations of ```checksysodesol```. I want to discuss with you about that and if required, it would be best open an issue for it.

Reply:
Best to open a new issue for problems with checksysodesol.

I think that the Piecewise solutions should be removed for now because they are poorly implemented (see various issues with dsolve) and they are not done for any of the other solvers including solve, solveset, the other dsolve solvers etc. Instead only the generic case should be returned.

The different piecewise cases are correct but they are not something that can be applied consistently across dsolve. The number of possibilities scales exponentially as the size of the system increases. The support for this would have to built from the bottom up through solve, eigenvalues etc. Then we could build on that in a consistent way in dsolve. I have discussed ideas along these lines in e.g. #16861

The Piecewise here would only need to have 2 cases (the 3rd is redundant) to cover the two possibilities for the Jordan blocks of a 2x2:
```
[[lam1, 0], [lam2, 0]]
[[lam, 1], [lam, 0]]
```